includes/amp-helper-functions.php

```diff
@@ -301,6 +301,68 @@ function amp_add_frontend_actions() {
 	add_action( 'wp_head', 'amp_add_amphtml_link' );
 }
 
+/**
+ * Determine whether AMP is available for the current URL.
+ *
+ * @todo This should go into the AmpProject\AmpWP namespace?
+ * @since 1.6
+ *
+ * @return bool True if there is an AMP version, false otherwise.
+ */
+function is_amp_available() {
+	global $wp_query;
+
+	if ( ! $wp_query instanceof WP_Query || ! did_action( 'wp' ) ) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			esc_html(
+				sprintf(
+					/* translators: %1$s: function name, %2$s: the current action, %3$s: the wp action, %4$s: the WP_Query class, %5$s: the amp_skip_post() function */
+					__( '%1$s was called too early and so it will not work properly. WordPress is currently doing the "%2$s" action. Calling this function before the "%3$s" action means it will not have access to %4$s and the queried object to determine if it is an AMP response, thus neither the "%5$s" filter nor the AMP enabled toggle will be considered.', 'amp' ),
+					__FUNCTION__ . '()',
+					current_action(),
+					'wp',
+					'WP_Query',
+					'amp_skip_post()'
+				)
+			),
+			'1.6.0'
+		);
+		return false;
+	}
+
+	$queried_object = get_queried_object();
+	if ( current_theme_supports( AMP_Theme_Support::SLUG ) ) {
+		// Abort if in Transitional mode and AMP is not available for the URL.
+		$availability = AMP_Theme_Support::get_template_availability( $wp_query );
+
+		if ( ! $availability['supported'] ) {
+			return false;
+		}
+
+		// Check to see if there are known unaccepted validation errors for this URL.
+		$current_url       = amp_get_current_url();
+		$validation_errors = AMP_Validated_URL_Post_Type::get_invalid_url_validation_errors( $current_url, [ 'ignore_accepted' => true ] );
+		$error_count       = count( $validation_errors );
+
+		if ( $error_count > 0 ) {
+			return false;
+		}
+	}
+
+	if ( ! (
+		$queried_object instanceof WP_Post &&
+		$wp_query instanceof WP_Query &&
+		( $wp_query->is_singular() || $wp_query->is_posts_page ) &&
+		post_supports_amp( $queried_object ) )
+	) {
+		// Abort if in Reader mode and the post doesn't support AMP.
+		return false;
+	}
+
+	return true;
```

Comment:
As discussed we will probably need to filter this to allow plugins to force AMP to be enabled or disabled ona  per-URL basis. See #2817.

Should this filter allow overriding all the above `return false` statements as well so to force a response that would have been not AMP to be instead be forced to be AMP? This would end up being a way to implement #4626.

Reply:
No need to do that now. We can revisit this later.